### PR TITLE
Add flag to enable activation sharding

### PR DESCRIPTION
**Pull request by wonjoo-wj**
With this PR, you can add flag `--enable_activation_sharding True` to your command to enable activation sharding. By default, this is set to false.

### CODE REVIEW

File: llama/model.py [hunk 1]
```diff
@@ -39,6 +39,8 @@ class ModelArgs:
     max_seq_len: int = 2048
     quant: bool = False
 
+    enable_activation_sharding: bool = False
+
     def print_values(self):
         print(f'[WONJOO] ModelArgs')
```
Comment (yeounoh) [inline llama/model.py]: remove '[wonjoo]'

**Review by yeounoh — APPROVED** [review]
LGTM minor suggestion

File: llama/model.py [hunk 2]
```diff
@@ -203,12 +207,16 @@ def forward(
...
-        num_devices = xr.global_runtime_device_count()
-        device_ids = torch.arange(num_devices)
-        data_model_mesh = xs.Mesh(device_ids, (4, 1, 2))
-        xs.mark_sharding(output, data_model_mesh, (0, 1, 2), use_dynamo_custom_op=True)
+        # Activation output sharding
+        import torch_xla.experimental.dynamo_mark_sharding
+        if self.enable_activation_sharding:
+            # num_devices = 8  # xr.global_runtime_device_count()
+            # device_ids = [i for i in range(num_devices)]
+            device_ids = [0, 1, 2, 3, 4, 5, 6, 7]
```
Comment (yeounoh) [inline llama/model.py]: Ah. can we actually pass in the number of devices and have it generates this? or call `num_devices = xr.global_runtime_device_count()` when appropriate

Comment (yeounoh) [reply llama/model.py]: We should form the mesh accordingly, [n_devices//2, 1, 2]

Comment (wonjoo-wj) [reply llama/model.py]: Updated

**Comment by wonjoo-wj** [comment]
@yeounoh, I've updated the code to dynamically fetch the num_device, device_ids, mesh_shape by calculating num_device outside of the torch.compile'd function. Verified that now it properly accepts the argument `--enable_activation_sharding` like:
```
PJRT_DEVICE=TPU XLA_USE_SPMD=1 python3 example_text_completion.py --ckpt_dir checkpointing/7B/ --tokenizer_path spiece.model --max_seq_len 2048 --max_gen_len 1000 --max_batch_size 1 --mp False --dynamo True --spmd True --enable_activation_sharding True
```
I'll go ahead and merge this for now and follow-up if there are any other things to address.

**Pull request merged**